sycl/tools/pi-trace/pi_trace.cpp

```diff
@@ -0,0 +1,84 @@
+//==----------- pi_trace.cpp.cpp -------------------------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+/// \file pi_trace.cpp
+/// A sample XPTI subscriber to demonstrate how to collect PI function call
+/// arguments.
+
+#include "xpti_trace_framework.h"
+
+#include "pi_arguments_handler.hpp"
+
+#include <detail/plugin_printers.hpp>
+
+#include <iostream>
+#include <mutex>
+#include <string>
+#include <string_view>
+#include <thread>
+
+static uint8_t GStreamID = 0;
+std::mutex GIOMutex;
+
+sycl::xpti_helpers::PiArgumentsHandler ArgHandler;
+
+// The lone callback function we are going to use to demonstrate how to attach
+// the collector to the running executable
+XPTI_CALLBACK_API void tpCallback(uint16_t trace_type,
+                                  xpti::trace_event_data_t *parent,
+                                  xpti::trace_event_data_t *event,
+                                  uint64_t instance, const void *user_data);
+
+// Based on the documentation, every subscriber MUST implement the
+// xptiTraceInit() and xptiTraceFinish() APIs for their subscriber collector to
+// be loaded successfully.
+XPTI_CALLBACK_API void xptiTraceInit(unsigned int major_version,
+                                     unsigned int minor_version,
+                                     const char *version_str,
+                                     const char *stream_name) {
+  if (std::string_view(stream_name) == "sycl.pi.arg") {
+    GStreamID = xptiRegisterStream(stream_name);
+    xptiRegisterCallback(
+        GStreamID, (uint16_t)xpti::trace_point_type_t::function_with_args_begin,
+        tpCallback);
+    xptiRegisterCallback(
+        GStreamID, (uint16_t)xpti::trace_point_type_t::function_with_args_end,
+        tpCallback);
+
+#define _PI_API(api, ...)                                                      \
+  ArgHandler.set##_##api([](auto &&... Args) {                                 \
+    std::cout << "---> " << #api << "("                                        \
+              << "\n";                                                         \
+    sycl::detail::pi::printArgs(Args...);                                      \
```

Comment:
Not sure I fully understand this code. It seems like this lambda will be called with a single argument - tuple. It doesn't seem like `printArgs` can print tuples. Am I missing something?

Also, can we have a simple test? It would help with understanding how it works.

Reply:
I use [`std::apply`](https://en.cppreference.com/w/cpp/utility/apply) to unpack tuple into separate function call arguments inside `handle(void *)`.